app/src/main/kotlin/eu/kanade/tachiyomi/ui/reader/loader/PageLoader.kt

```diff
@@ -58,79 +69,103 @@ abstract class PageLoader {
+            onRequest(realIndex)
         }
 
         // Prefetch to disk
-        val prefetchRange = if (index >= lastRequestIndex) {
-            index + 1..(index + prefetchPageCount).coerceAtMost(size - 1)
+        val prefetchRange = if (visualIndex >= lastRequestIndex) {
+            visualIndex + 1..(visualIndex + prefetchPageCount).coerceAtMost(size - 1)
         } else {
-            index - 1 downTo (index - prefetchPageCount).coerceAtLeast(0)
+            visualIndex - 1 downTo (visualIndex - prefetchPageCount).coerceAtLeast(0)
         }
-        val pagesAbsent = prefetchRange.filter { pages[it].status.value == Page.State.QUEUE }
+        val pagesAbsent = prefetchRange.map { stateList[it] }.filter { it.status.value == Page.State.QUEUE }
+
         val start = if (prefetchRange.step > 0) prefetchRange.first else prefetchRange.last
         val end = if (prefetchRange.step > 0) prefetchRange.last else prefetchRange.first
-        prefetchPages(pagesAbsent, start - 5 to end + 5)
+        prefetchPages(pagesAbsent.map { it.index }, start - 5 to end + 5)
 
         // Prefetch to memory
-        val range = index - 3..index + 3
-        pagesAbsent.forEach {
-            if (it in range) onRequest(it)
+        val range = visualIndex - 3..visualIndex + 3
+        pagesAbsent.forEach { absentPage ->
+            if (stateList.indexOf(absentPage) in range) onRequest(absentPage.index)
         }
 
-        lastRequestIndex = index
+        lastRequestIndex = visualIndex
     }
 
-    fun retryPage(index: Int, orgImg: Boolean = false) {
+    fun retryPage(page: ReaderPage, orgImg: Boolean = false) {
+        val index = page.index
         notifyPageWait(index)
         onForceRequest(index, orgImg)
     }
 
+    private val Int.page
+        get() = trackMap[this]!!
+
+    // Following accepts real index!!!
+
     protected abstract fun prefetchPages(pages: List<Int>, bounds: Pair<Int, Int>)
 
     protected abstract fun onRequest(index: Int)
 
     protected abstract fun onForceRequest(index: Int, orgImg: Boolean)
 
-    fun cancelRequest(index: Int) {
-        onCancelRequest(index)
-    }
-
-    protected abstract fun onCancelRequest(index: Int)
-
     fun notifyPageWait(index: Int) {
-        pages[index].reset()
+        index.page.reset()
     }
 
     fun notifyPagePercent(index: Int, percent: Float) {
-        pages[index].progress = (percent * 100).toInt()
-        pages[index].status.compareAndSet(Page.State.QUEUE, Page.State.DOWNLOAD_IMAGE)
+        val page = index.page
+        page.progress = (percent * 100).toInt()
+        page.status.compareAndSet(Page.State.QUEUE, Page.State.DOWNLOAD_IMAGE)
     }
 
     fun notifyPageSucceed(index: Int, image: Image, replaceCache: Boolean = true) {
-        if (replaceCache) {
-            cache.put(index, image)
+        val page = index.page
+        if (image.hasQRCode) {
+            println("$index has QRCode")
+            stateList.remove(page)
```

Comment:
I'm not sure I like the idea of modifying the list.
It will cause the total number to change during reading, which is kinda weird.
Maybe we can add a new page state like "Blocked" or something, and cover the image with a mask (blur or just black), which has a "Show anyway" button in case of false positive.

Reply:
Sounds good, make pages size mutable & observable introduce much maintain burden

Reply:
Done